Reject sent header blocks with bad :authority/Host headers

alexwlchan opened the pull request:
Quick something I cooked up in my lunch break.

This continues the work started in #252. Intention is that if the user tries to send :authority or Host headers that don't conform, we should raise an `InvalidHeaderBlock` exception.

For now, I've just refactored the existing send_headers pipeline into `utilities.py`, ready to add the new logic. I'll look at the rest of it this evening/tomorrow.

alexwlchan commented on h2/utilities.py:
```diff
@@ -41,7 +41,7 @@
 
 
-def secure_headers(headers):
+def secure_headers(headers, hdr_validation_flags):
```
This is a public function, although the only place it was used was in `stream.py`.

Is it a problem that the argument signature is changing? Or could we add an underscore and make it non-public, to match the other header pipeline functions in this file?

Lukasa replied:
It's not a public function. =)

hyper-h2 is very clear in its documentation: publicness is not signalled by underscores per-se, but by what is documented in the API docs. This is not documented in those docs (in fact, the entire `utilities` module is not). Therefore, we can do whatever we want here.

alexwlchan commented on h2/utilities.py:
```diff
@@ -160,6 +160,7 @@ def validate_headers(headers, hdr_validation_flags):
     """
     Validates a header sequence against a set of constraints from RFC 7540.
 
+    :param headers: An iterable of headers.
```
Some naughty person wrote an incomplete argument list in the docstring, tut tut.

alexwlchan commented on h2/utilities.py:
```diff
@@ -267,9 +243,59 @@ def _reject_pseudo_header_fields(headers, hdr_validation_flags):
+    # It is an error for a request header block to contain neither
+    # an :authority header nor a Host header.
+    if not authority_present and not host_present:
+        if mode == 'receiving':
```
I’m not super in love with adding ~another level of `if` statements to this function. Do we need the extra exception type?

Any suggestions for making this cleaner?

Lukasa replied:
Sorry, this got out of hand a bit.

I suggested in the prior PR that you pull the core of this validation logic out to a separate function, leaving just the checks in here and then a call to that function. That would resolve this problem for you: we can then have one function for receiving and one for sending, each with different checks, and then the sending function can simply wrap the error from the receiving one.

alexwlchan replied:
So I ~think you’re suggesting a function something like:

    def core_validating_logic(authority, host):
        # check at least one is set
        # if both set, check authority == host

which raises a ProtocolError if it fails.

Then the function for sending would catch the `ProtocolError`, an re-raise as an `InvalidHeaderBlockError` with the same exception message. And we tweak the error messages so they don’t refer to receiving/sending, and the same error text can apply to both.

Have I got that right?

Lukasa replied:
That can do it, or we can even pass the exception class in to that function and have it raise from that. But yeah, that's basically right.

alexwlchan commented on h2/utilities.py:
```diff
@@ -298,22 +324,94 @@ def _check_host_authority_header(headers, hdr_validation_flags):
-                "Received mismatched :authority and Host headers: %r / %r" %
-                (authority_header_val, host_header_val)
-            )
+def _secure_headers(headers, hdr_validation_flags):
```
I moved this down just so all the sent-headers pipeline functions are nicely grouped together.

Lukasa replied:
Can we undo that and consider it a refactor for a separate PR? It makes it hard to confirm that there aren't function changes. =)

alexwlchan commented:
So I’ve pulled out the header validation, and currently it gets called in `H2Stream._build_headers_frames`, where the original send-header-normalization logic lived. This is causing a whole bunch of tests to fail, which I can sort out in a bit. First let’s get the core function right.

Having now dug into this a bit, a few thoughts:

1. Is this the right place to be putting it? I wonder if `H2Stream.send_headers` is a better place for it – that’s where we raise other errors with the headers we’re about to send, e.g. sending trailers without END_STREAM set.
2. The current header validation errors in `H2Stream.send_headers` all raise `ProtocolError` when you try to send something invalid. Do we need a special `InvalidHeaderBlockError`?
3. There are other checks in the received header validation pipeline that we don’t enforce on outgoing headers (e.g. the Connection header shouldn’t be present in a header block). If I’ve understood correctly, you could have hyper on both ends of a connection, and client-hyper could send a header block that server-hyper would reject. That seems a bit odd.

Lukasa commented:
1. So the reason it's in _build_header_frames is if it applies to sending all possible headers frames, including PUSH_PROMISE. That's an easy trap to fall into when we put it elsewhere: the headers for PUSH_PROMISEs also need validation. In fact, we probably overlooked this in #252.
2. ProtocolError is fine.
3. It's definitely odd. I'm open to fixing that up in some cases. =)

alexwlchan commented:
1. Okay, so next question: within `build_headers_frames`, how do we know if we’re in a trailer frame? Looking at `send_headers()`, it looks like maybe we want to be looking at `self.state_machine.trailers_send`, and if that’s `True`, we’re on a trailer?

2. Switched to ProtocolError.

3. I’ll raise a separate issue to track that weirdness, so that discussion about which header checks we might want to apply doesn’t clutter up this thread.

alexwlchan commented:
Other Q: is there an equivalent of `_build_header_frames` for received headers, that we can apply to the work done in #252?

Lukasa commented:
@alexwlchan 

1. Nah, in this instance it would be better to emit an internal-only event from the state machine that we suppress and do not emit, then use that to detect. Given that the state machine knows, we should get it to tell us.

alexwlchan commented:
In the two places we use `_build_headers_frames`, we ask the state machine to process input, but then assert we don’t get any events back. So would it be sensible to emit the internal-only event here, and pass that event into `_build_headers_frames`? This is all internal, so changing the argument signatures shouldn’t be (too) much of a problem.

Lukasa commented:
Yes, that's absolutely fine. =)

alexwlchan commented on h2/events.py:
```diff
@@ -136,6 +136,51 @@ def __repr__(self):
         )
 
 
+class ResponseSent(object):
```
These are just ResponseReceived and TrailerReceived redressed with different names. I omitted a few of the extra fields, because we don’t have the logic to fill them in. (Unless I missed something.) Was that right?

Lukasa replied:
So I think we'd like to clarify in some way that these are internal events. It's probably enough to put appropriate docstrings on them. Otherwise this is fine.

Out of interest, why don't we need RequestSent?

Lukasa replied:
If the answer is that we probably do, can we replace both with a singular `HeadersSent`? When `TrailersSent` exists, I think those two names are clear enough.

alexwlchan commented on h2/utilities.py:
```diff
@@ -291,29 +295,101 @@ def _check_host_authority_header(headers, hdr_validation_flags):
 
     for header in headers:
-        if header[0] == b':authority':
+        if header[0] in (b':authority', u':authority'):
```
I’m sneaking this change in here, because it slipped through in #252, and it caused some tests to fail when I was playing with the sent header blocks. Okay? This mirrors what we do in `authority_from_headers`.

Lukasa replied:
This shouldn't actually be needed here because received header blocks should always be bytes: they only get optionally decoded after validation. However, given that this loop is now basically identical to the one in `_check_send_authority_host_header`, I recommend hoisting this loop out to `_validate_host_authority_header` too. Let's duplicate as little code as possible.

alexwlchan replied:
Done.

alexwlchan commented:
First stab at this pushed. Once again broken a lot of the tests – I’m guessing I’ve bumped a stream ID somewhere? I’m not quite sure what I’ve broken, so I’m going to leave this for tonight.

Lukasa commented on h2/stream.py:
```diff
@@ -1018,14 +1018,22 @@ def stream_reset(self, frame):
+        try:
+            is_trailer = isinstance(events[0], TrailersSent)
+        except IndexError:
+            is_trailer = False
```
Can we have a blank line after this block?

alexwlchan replied:
Done.

Lukasa replied:
That would be ideal. =)

Lukasa commented on h2/utilities.py:
```diff
@@ -291,29 +295,101 @@ def _check_host_authority_header(headers, hdr_validation_flags):
+        authority_header_val, host_header_val)
+
+
+def validate_sent_headers(headers, hdr_validation_flags):
```
I think this function needs a better name. Alternatively, we should normalize *first*, and then validate. Keeping `_lowercase_header_names` as a separate method that applies before the validation step doesn't seem unreasonable to me.

Lukasa commented on h2/events.py:
```diff
@@ -136,6 +136,56 @@ def __repr__(self):
         )
 
 
+class HeadersSent(object):
```
Part of me wonders if we should prefix these class names with underscores, just to make *really sure* we don't get confused.

Lukasa commented on h2/events.py:
```diff
@@ -136,6 +136,56 @@ def __repr__(self):
+    This is an internal event, used to distinguish between headers
+    and trailers.
+
+    .. versionadded:: 2.5.0
```
We don't need a versionadded marker here: that's only necessary for public classes.

Lukasa replied:
Indeed, it's at real risk of being wrong as I may well backport this change.

alexwlchan commented:
Applied markups and rebased against master.

Lukasa commented on h2/stream.py:
```diff
@@ -787,15 +789,14 @@ def push_stream_in_band(self, related_stream_id, headers, encoder):
         # Because encoding headers makes an irreversible change to the header
         # compression context, we make the state transition *first*.
 
-        # This does not trigger any events.
```
You removed this comment, but it seems to still be accurate.

Lukasa replied:
Note that push promise is going to put some tricky assumptions to you: in particular, the client/server dichotomy no longer applies quite so cleanly. The headers in a push-promise frame are effectively *client* headers sent *by* a server.

Lukasa replied:
That may mean we need a third event here, which you can use to help calculate your state tuple.

Lukasa commented on h2/utilities.py:
```diff
@@ -279,33 +280,30 @@ def _reject_pseudo_header_fields(headers, hdr_validation_flags):
+    a trailer, check that:
+     1. At least one of these headers is set.
+     2. If both headers are set, they match.
+    Raises ``exception_class`` if the header block is invalid.
```
This comment is present but not acted on.

Lukasa commented on h2/utilities.py:
```diff
@@ -279,33 +280,30 @@ def _reject_pseudo_header_fields(headers, hdr_validation_flags):
     # enforced by the _reject_pseudo_header_fields() pipeline.
     #
     # TODO: We should also guard against receiving duplicate Host headers.
+    # TODO: We should guard against sending duplicate headers.
```
We probably don't need this #TODO twice. ;)

Lukasa commented on h2/utilities.py:
```diff
@@ -319,13 +317,82 @@ def _check_host_authority_header(headers, hdr_validation_flags):
+    if hdr_validation_flags.is_client or hdr_validation_flags.is_trailer:
+        return headers
+
+    return _validate_host_authority_header(headers)
```
Shiny, nice change! I had some vague concerns about what we were going to do in the absence of `yield from`, but changing this function to always return a generator without having any `yield` statements in it is a great solution to that problem. Well done!

Lukasa commented:
Alright, fab. Aside from the problems with pushes and a few minor style notes, this is getting much better!

We'll want to add a chunk of tests, much like the ones we have for the current validation pipeline, to confirm that we now forbid emitting the same headers we've been rejecting. This should be easy enough to do, though, because you just need to fire them into the connection as the first client request and watch it explode. =)

alexwlchan commented on h2/stream.py:
```diff
@@ -787,15 +805,14 @@ def push_stream_in_band(self, related_stream_id, headers, encoder):
             StreamInputs.SEND_PUSH_PROMISE
         )
-        assert not events
+        print("@@AWLC events is %r" % events)
```
Clearly this debugging line should come out before this gets merged. (Oops!)

alexwlchan commented:
@Lukasa I'll continue making markups and try to get the existing tests passing when I get home from work.

Any ideas why I've broken so much of the header indexing? e.g. https://travis-ci.org/python-hyper/hyper-h2/jobs/146619142#L332

Lukasa commented on h2/stream.py:
```diff
@@ -216,23 +224,31 @@ def send_new_pushed_stream(self, previous_state):
+
+        event = _PushPromiseSent()
+        event.stream_id = self.stream_id
+        return [event]
```
This doesn't seem right. When we receive a push promise we spin *two* state machine methods. Conceptually, one of them owns the pushed headers and the other one does not.

We don't need a `PushPromiseSent` event in this case, because we *receive* a `PushPromiseReceived` event.

Lukasa commented on h2/stream.py:
```diff
@@ -132,7 +133,12 @@ def request_sent(self, previous_state):
-        return []
+
+        # TODO: Is it correct that we'd never send trailers on a response,
+        # or have I misread RFC 7230?
```
You've misread it: you can send trailers whenever you want.

Lukasa commented:
Hey @alexwlchan, what's the status here?

alexwlchan commented:
Sorry @Lukasa, last week was tricky and I didn’t have time for OSS stuff. Hoping to get back to it in the next day or so.

Lukasa commented:
That's ok, there's no rush. =)

alexwlchan commented:
Made a few markups, now I need to sort out the internal events that we fire (in particular, I now have a `_PushPromiseSent()` event but never actually use it). Looking through the sending methods on `H2StreamStateMachine`, this is what I think I should be doing:

* `request_sent() / response_sent()`, the logic should be:

   - if we've never sent headers, emit `_HeadersSent()`
   - if we’ve already sent headers, this is a trailer, so emit `_TrailersSent()`

  This is already implemented for `response_sent()`.

  (Also, can a server send requests?)

* `send_new_pushed_stream()` -- per your comment, we do nothing, because we receive a `PushPromiseReceived`

* `send_push_promise()` -- we should emit a `_PushPromiseSent` event.

* `send_informational_response()` and `send_alt_svc()` -- not sure what I should be doing here.

Plus tidying up all the places we assert we don‘t receive events, and updating the state tuples.

@Lukasa Does that sound sensible, or have I missed/misunderstood something?

Lukasa commented:
This sounds generally sensible. Let me elaborate on some points, because they touch on a series of bits of trickiness that I think the code doesn't necessarily make entirely clear (sadly).

1. Can a server send requests? Yes, but they are *pushed* requests and are contained entirely in a PUSH_PROMISE frame. No HEADERS frame a server emits can ever be for a request.
2. `send_new_pushed_stream` does not lead to an event being emitted. Prior to this change, no sending events methods emit events.
3. `send_new_pushed_stream` vs `send_push_promise`. We don't actually need both of these to emit an internal event. 

    The reason for this can be seen in `H2Connection.push_stream`. In short, though: a PUSH_PROMISE frame actually causes *two* stream-level state transitions. The first is for the stream on which the PUSH_PROMISE frame is sent (logically the *parent* stream). This stream has `H2Stream.push_stream_in_band` called on it, leading to `send_push_promise` being called, which currently returns no event. The second is for the stream that the PUSH_PROMISE *creates* (logically the *child* stream). This stream has `H2Stream.locally_pushed` called on it, which leads to `send_new_pushed_stream` being called on it and emitting the `PushPromiseReceived` event.

    This is a very long-winded way of saying: only *one* of these two methods should to return an event that we search for. More importantly, though: we don't need them to! Because, when we push a stream, these are *always* request headers. They cannot be trailers, they cannot be response headers, they must *always* be request headers. So we don't need an event at all: we know exactly what's going on here.
4. You don't need to worry about ALT_SVC, it has nothing to do with headers.
5. Informational responses are 1XX responses. They should be treated as ordinary response headers, though in the future we may want to validate them separately as they technically have different rules (e.g. no content-length allowed).

Does that clarify?

alexwlchan commented:
Okay, I think that all makes sense. I think the latest commit emits the correct events based on what you’ve said – could you double-check please? Once that’s done, I can start writing tests to check we reject the correct header set.

I’m also confused by the problem in the header indexing tests; I’m going to play around with git bisect to see where that started happening.

alexwlchan commented:
Found the header indexing bug: at some point I swapped the order in which we normalise sent headers. It was originally _lowercase, then secure_, and I’d turned it into _secure, then lowercase_.

alexwlchan commented on h2/events.py:
```diff
@@ -136,6 +136,52 @@ def __repr__(self):
+        self.stream_id = None
+
+        #: The headers.
+        self.headers = None
```
I just copied this field from some of the other internal events. But I can’t actually see where it gets populated on, say, `HeadersReceived`. Do we actually need it for these internal events, given that all we care about is whether this is a trailer or not?

Lukasa replied:
This ordinarily gets populated by the stream object itself, because the state machine doesn't get access to the headers. We don't need it here. =) Nor do we even really need the stream ID.

Lukasa commented on h2/events.py:
```diff
@@ -136,6 +136,52 @@ def __repr__(self):
+        self.stream_id = None
+
+        #: The trailers themselves.
+        self.headers = None
```
Similarly we don't need these.

Lukasa commented on h2/stream.py:
```diff
@@ -131,8 +132,13 @@ def request_sent(self, previous_state):
+            self.headers_sent = True
+            event = _HeadersSent()
+        else:
+            event = _TrailersSent()
```
Are you sure this if branch is necessary? If it is, I think we need `self.trailers_sent` to be populated as well.

Lukasa replied:
uh...actually, I think this branch is un-hittable. It looks a bit like maybe clients can't send trailers at all. That might be a separate bug.

Lukasa replied:
Ah, no, they can, but it happens in `response_sent`. So I'm right, this branch is unhittable and we don't need it.

Lukasa commented on h2/stream.py:
```diff
@@ -1023,14 +1030,24 @@ def stream_reset(self, frame):
                               encoder,
-                              first_frame):
+                              first_frame,
+                              events):
```
Rather than pass the events in here, can we hoist the building of the validation flags up a level? Maybe we need to write a helper function to do that, but that shouldn't be hard, and it avoids this method doing too much.

Lukasa commented:
Ok, I've got a few smaller notes about an unreachable branch and some other stuff, but this is looking broadly right now. =D Great work so far!

alexwlchan commented on h2/events.py:
```diff
@@ -136,6 +136,52 @@ def __repr__(self):
 
+class _HeadersSent(object):
+    """
+    The _HeadersSent event is fired whenever headers are received.  This event
```
> whenever headers are received

No.

alexwlchan commented on test/test_invalid_headers.py:
```diff
@@ -95,30 +97,49 @@ class TestFilter(object):
-            assert headers == h2.utilities.validate_headers(
-                headers, hdr_validation_flags)
+            assert headers == list(validation_function(
+                headers, hdr_validation_flags))
```
Possibly this is the wrong fix; this was an attempt to fix a failure when a client tries to send an empty list of headers which aren't trailers:

```
        try:
>           assert headers == validation_function(
                headers, hdr_validation_flags)
E               assert [] == <generator object _validate_...thority_header at 0x279b910>
E                 Full diff:
E                 - []
E                 + <generator object _validate_host_authority_header at 0x279b910>

test/test_invalid_headers.py:127: AssertionError
------------------------------------------------------------------------- Hypothesis -------------------------------------------------------------------------
Falsifying example: test_range_of_acceptable_outputs(self=<test_invalid_headers.TestFilter at 0x28f0510>, headers=[], validation_function=validate_sent_headers, is_client=True, is_trailer=False)
============================================================ 1 failed, 31 passed in 3.04 seconds =============================================================
```

alexwlchan commented on test/test_invalid_headers.py:
```diff
@@ -95,30 +97,49 @@ class TestFilter(object):
+    validator_combinations = list(itertools.product(
+        validation_functions,
+        (True, False),
+        (True, False),
```
Notes to self:

1. Do you definitely need to cast to a list?
2. Better if this had the logic for constructing header validation flags, and those were passed directly into the tests; hoist a bit more logic out of the tests themselves.

alexwlchan commented:
Notes for me before I go back to work (thinking aloud more than anything else):

* How does this interact with the config options for skipping this validation?
* Eventually we'd like to reject all the invalid header blocks in `test_invalid_headers.py` when we try to send them, for now we'll need to skip at least some of them.
* There was talk of writing some documentation for this change. Write some!

alexwlchan commented:
Okay, final update from me for the weekend. Comments/questions for @Lukasa:

* We need a flag to skip this header validation for people who like to live on the edge. Do we need a generic "skip all outbound header validation" flag, or should we break it down into more granular flags for each check we perform? I lean towards the former.

  Once I know what this flag looks like, I can update the docs.

* Tests are failing because we don’t have 100% coverage of `authority_from_headers()` in the utilities module – specifically, the `return None` if we don’t find an :authority header. Because any headers we try to send without an :authority header are being rejected before we hit this function. 😄 

  Probably the correct thing here is to test the config object really will let you send invalid headers, including one set without an :authority header.

Lukasa commented:
> Do we need a generic "skip all outbound header validation" flag, or should we break it down into more granular flags for each check we perform? I lean towards the former.

The former, for sure. Probably we need two flags, `validate_outbound_headers` (defaults to `True`) and `normalize_outbound_headers` (defaults to `True`), but for this PR feel free to just implement the first.

We obviously also want symmetrical flags for inbound headers.

> Probably the correct thing here is to test the config object really will let you send invalid headers, including one set without an :authority header.

Yup, sounds good.

alexwlchan commented:
I’ve squashed some of the earlier commits, and rebased against master.

1. The latest commit adds three of the four config options suggested in your previous comment. I haven‘t added a `normalize_outbound_headers` flag because, AFAICT, that’s not something we ever do, so it can’t be turned off. Or have I missed something here?

2. The proliferation of boolean config options – all using very similar code – feels like a code smell. I’m tempted to wrap it all up in some sort of `BooleanConfigOption` descriptor, and cut out the repetitive code – but it feels a bit like I’m just wielding a descriptor-shaped hammer. What do you think?

3. I’m trying to work out where we should actually make use of these settings. The `H2Connection` class owns the config, but it’s `H2Stream` that actually calls the validation/normalization functions. I’m struggling to think of a good way to pass the config options down into `H2Stream`; unless there’s a way for `H2Stream` to peek "up" at its owner and read the config from there?

Lukasa commented:
> I haven‘t added a normalize_outbound_headers flag because, AFAICT, that’s not something we ever do, so it can’t be turned off.

I presume you mean *inbound* headers? In which case, that's correct. =)

> The proliferation of boolean config options – all using very similar code – feels like a code smell. I’m tempted to wrap it all up in some sort of BooleanConfigOption descriptor, and cut out the repetitive code – but it feels a bit like I’m just wielding a descriptor-shaped hammer. What do you think?

Let's open a new issue to track this, and then deal with it as a refactor in a separate PR. I'm happy to remove the extra code, just want to try to reduce scope creep in this PR. =)

> I’m trying to work out where we should actually make use of these settings. The H2Connection class owns the config, but it’s H2Stream that actually calls the validation/normalization functions. I’m struggling to think of a good way to pass the config options down into H2Stream; unless there’s a way for H2Stream to peek "up" at its owner and read the config from there?

We should pass the config object into the `H2Stream` on construction.

alexwlchan commented:
Oh so nearly there. We’re now passing the `config` object to the `H2Stream`, skipping the validation/normalization where appropriate, and have some tests for the new function. Shiny!

Tests are failing because of the issue fixed in #285; once that’s merged, I’ll rebase atop that. Applying the fix locally gets the tests passing.

The one last problem is the `normalize_sent_headers` parameter. I’m struggling to think of a good test for it, and without that test coverage is <100%. Ideally we’d verify that sent headers don’t get modified when you disable this setting, but AFAICT you can’t see what headers get sent after you’ve called `send_headers` – which seems perfectly sensible. Any ideas?

Lukasa commented:
@alexwlchan Feel free to use the `hpack` module to import a HPACK decoder that can decode the HEADERS frame we emit. =)

alexwlchan commented:
At last a green tick, hurrah! I believe this is now ready to merge – at the very least, there’s nothing else on my todo list for it.

The next few days are a bit busier, so progress might slow down again a bit. (Sorry, the last few weeks have been a bit up-in-the-air – obviously the best time to start a big patch. 😉 Hopefully the next few patches will be a bit smaller.)

Lukasa commented:
Heh, thanks @alexwlchan for all this work! I'll do a deep-dive review shortly, but I'm really pleased with how this patch has been shaping up.

Lukasa commented on h2/connection.py:
```diff
@@ -512,6 +512,7 @@ def _begin_new_stream(self, stream_id, allowed_ids):
             self.remote_settings.initial_window_size
         )
         s.inbound_flow_control_window = self.local_settings.initial_window_size
+        s.config = self.config
```
This looks like it just gets casually patched onto the `H2Stream`. Can we update `H2Stream.__init__` to create this member variable initialized to `None`?

Lukasa commented on h2/config.py:
```diff
@@ -77,3 +106,61 @@ def header_encoding(self, value):
         self._header_encoding = value
+
+    @property
+    def validate_sent_headers(self):
```
Minor nit, but it would be nicer to rename this to `validate_outbound_headers` for symmetry with `validate_inbound_headers`. Alternatively, swap to `sent`/`received`.

alexwlchan replied:
I'll go for `outbound` -- that's more accurate, as we validate the headers when they're outbound, not necessarily sent.

Lukasa commented on h2/stream.py:
```diff
@@ -684,6 +691,9 @@ def __init__(self, stream_id):
         self._authority = None
 
+        # The configuration for this stream.
+        self.config = H2Configuration()
```
Don't create a brand new config object. In fact, ignore my prior comment: we should probably provide this on the constructor, as we cannot reasonably expect streams to exist without it.

Lukasa commented on h2/stream.py:
```diff
@@ -754,12 +764,12 @@ def send_headers(self, headers, encoder, end_stream=False):
-        frames = self._build_headers_frames(headers, encoder, hf)
+        hdr_validation_flags = self._build_hdr_validation_flags(events)
+        frames = self._build_headers_frames(
+            headers, encoder, hf, hdr_validation_flags)
```
Sorry, the general project style for long lines like this is:

```python
frames = self._build_headers_frames(
    headers, encoder, hf, hdr_validation_flags
)
```

Lukasa commented on h2/stream.py:
```diff
@@ -795,7 +805,9 @@ def push_stream_in_band(self, related_stream_id, headers, encoder):
-        frames = self._build_headers_frames(headers, encoder, ppf)
+        hdr_validation_flags = self._build_hdr_validation_flags(events)
+        frames = self._build_headers_frames(
+            headers, encoder, ppf, hdr_validation_flags)
```
Same here.

Lukasa commented on h2/stream.py:
```diff
@@ -918,11 +930,12 @@ def receive_headers(self, headers, end_stream, header_encoding):
+            hdr_validation_flags = HeaderValidationFlags(
+                is_client=self.state_machine.client,
+                is_trailer=isinstance(events[0], TrailersReceived)
+            )
```
This can also use `_build_hdr_validation_flags` if we slightly broaden that method.

Lukasa commented on h2/stream.py:
```diff
@@ -1020,17 +1033,36 @@ def stream_reset(self, frame):
+        and validating header blocks.
+        """
+        try:
+            is_trailer = isinstance(events[0], _TrailersSent)
```
Let's add `TrailersReceived` to the list of trailery events, because we can then use this method in `receive_headers` too.

Lukasa commented on h2/utilities.py:
```diff
@@ -319,13 +317,82 @@ def _check_host_authority_header(headers, hdr_validation_flags):
-                "Received mismatched :authority and Host headers: %r / %r" %
-                (authority_header_val, host_header_val)
+                "Request header block must have matching :authority and "
+                "Host headers: %r / %r"
```
Let's change this to a statement of what is wrong: `"Request header block has mismatched :authority and Host headers"`

Lukasa commented:
Awesome, this is looking really good! I have left some inline notes, but generally speaking this is looking pretty shiny.

alexwlchan commented:
Ding. Review markups made. And a mere 77 comments later, are we done? :smile:

Lukasa commented on h2/stream.py:
```diff
@@ -1020,17 +1031,42 @@ def stream_reset(self, frame):
+                events[0], (_TrailersSent, TrailersReceived)
+            )
+        except IndexError:
+            is_trailer = False
```
Hrm. What's this fallback for, and when do we hit it?

alexwlchan replied:
There are cases where we don't emit an internal event &ndash; push promise is the example that springs to mind &ndash; and so `events` is empty. It's the places where we still have `assert not events`.

Lukasa replied:
Aha, ok. This feels like an easy source of bugs, but for now let's put a comment here.

alexwlchan replied:
Comment added, along with a TODO to make this unnecessary in future.

Lukasa commented:
I have one question, and then we should be safe.

Lukasa commented:
\o/ Thanks @alexwlchan, this turned into a marathon patch, but it looks really good! :sparkles:
